### PR TITLE
Add continue construct analysis to struct cfg analysis

### PR DESCRIPTION
**Pull request by s-perron**
Add the ability to identify which blocks are in the continue construct for a
loop, and to get functions that are called from those blocks, directly or
indirectly.

Part of https://github.com/KhronosGroup/SPIRV-Tools/issues/2912.

### CODE REVIEW

File: source/opt/struct_cfg_analysis.h [hunk 1]
```diff
@@ -103,6 +119,7 @@ class StructuredCFGAnalysis {
     uint32_t containing_construct;
     uint32_t containing_loop;
     uint32_t containing_switch;
+    bool in_continue;
```
Comment (ehsannas) [inline source/opt/struct_cfg_analysis.h]: The comment hasn't been updated for `|containing_switch|` and `|in_continue|`

File: source/opt/struct_cfg_analysis.cpp [hunk 2]
```diff
@@ -146,9 +161,59 @@ bool StructuredCFGAnalysis::IsContinueBlock(uint32_t bb_id) {
...
+    }
+  }
+
+  // Now collect all of the functions that are indirectly called as well.
+  while (!funcs_to_process.empty()) {
+    uint32_t func_id = funcs_to_process.front();
+    funcs_to_process.pop();
+    Function* func = context_->GetFunction(func_id);
+    if (called_from_continue.insert(func_id).second) {
+      context_->AddCalls(func, &funcs_to_process);
```
Comment (ehsannas) [inline source/opt/struct_cfg_analysis.cpp]: Does `AddCalls` go recursively deep into functions called by functions inside `func` (i.e. functions indirectly called by `func`) ?

File: source/opt/struct_cfg_analysis.cpp [hunk 3]
```diff
@@ -146,9 +161,59 @@ bool StructuredCFGAnalysis::IsContinueBlock(uint32_t bb_id) {
...
+  }
+  return it->second.in_continue;
+}
+
+bool StructuredCFGAnalysis::IsInContinueConstruct(uint32_t bb_id) {
+  while (bb_id != 0) {
+    if (IsInContainingLoopsContinueConstruct(bb_id)) {
+      return true;
+    }
+    bb_id = ContainingLoop(bb_id);
```
Comment (ehsannas) [inline source/opt/struct_cfg_analysis.cpp]: is this for a case where you have a loop inside a continue block?

File: test/opt/struct_cfg_analysis_test.cpp [hunk 4]
```diff
@@ -835,8 +1036,334 @@ OpFunctionEnd
...
+         %15 = OpFunctionCall %void %16
+               OpReturn
+               OpFunctionEnd
+         %14 = OpFunction %void None %7
+         %17 = OpLabel
+         %19 = OpFunctionCall %void %16
+               OpReturn
+               OpFunctionEnd
+         %16 = OpFunction %void None %7
+         %18 = OpLabel
```
Comment (ehsannas) [inline test/opt/struct_cfg_analysis_test.cpp]: Can we have `%16` also call one more level? (e.g.` OpFunctionCall %void %19`), and expect
`UnorderedElementsAre(14u, 16u, 19u)` ?

File: source/opt/struct_cfg_analysis.cpp (hunk 3 above)
Comment (s-perron) [reply source/opt/struct_cfg_analysis.cpp]: Yes, that is correct.  It walking loop next checking if a loop is in the continue construct of the next loop.

File: source/opt/struct_cfg_analysis.cpp (hunk 2 above)
Comment (s-perron) [reply source/opt/struct_cfg_analysis.cpp]: `AddCalls` only add the functions called directly by `func`.  The comment for `AddCalls` could be clearer.

File: test/opt/struct_cfg_analysis_test.cpp (hunk 4 above)
Comment (s-perron) [reply test/opt/struct_cfg_analysis_test.cpp]: done.

File: source/opt/struct_cfg_analysis.h (hunk 1 above)
Comment (s-perron) [reply source/opt/struct_cfg_analysis.h]: Done

File: source/opt/struct_cfg_analysis.cpp (hunk 2 above)
Comment (s-perron) [reply source/opt/struct_cfg_analysis.cpp]: I updated the comments.

**Pull request merged**